projects/feature-activation/0001-feature-activation.md

```diff
@@ -0,0 +1,558 @@
+
+With that being said, it is possible that a feature may or may not be activated at the end of an activation process. To preserve flexibility, that behavior will be configurable on a per-feature basis through the `activate_on_timeout` attribute, explained below. That attribute is analogous to the infamous Bitcoin LOT attribute (lock in on timeout) [6], proposed in BIP 8.
+
+#### Burying feature activations
+[Burying feature activations]: #burying-feature-activations
+
+Burying a feature activation means changing code so "history is erased", and it's not clear that a feature was activated using the activation process.
+
+It is allowed to change code that is conditional on feature activation, but **NOT** to remove a feature from the definition, as described in the [Feature and Criteria definition] section.
+
+## Feature and Criteria definition
+[Feature and Criteria definition]: #feature-and-criteria-definition
+
+The definition of new features available for feature activation should be as simple as defining the feature name and the criteria required to evaluate its activation state. Once defined, ideally, it should not be changed. The `Feature` enum will contain all features, past and future, activated or not:
+
+```python
+class Feature(Enum):
+    NANO_CONTRACTS = Criteria(name='NANO_CONTRACTS', ...)
+    SOME_NEW_OP_CODE = Criteria(name='SOME_NEW_OP_CODE', ...)
+    SOME_OTHER_FEATURE = Criteria(name='SOME_OTHER_FEATURE', ...)
+```
+
+The `Criteria` data class is used to define attributes such as the feature name and activation requirements. Each name should be unique across all features, past and future. Its definition is provided in the [Criteria configuration] section.
+
+Different features with different criteria should be configurable separately for each network, mainnet and testnet. Feature definitions should **NEVER** be removed, even after activation, as explained in the [Burying feature activations] section. This is meant to preserve history.
+
+## Feature States
+[Feature States]: #feature-states
+
+For each block, there's a state associated with each feature. Possible states are similar to BIP 8, but `LOCKED_IN` and `MUST_SIGNAL` are left out and kept as future possibilities. These are the possible states:
+
+- `DEFINED`: Represents that a feature is defined. It's the first state for each feature. The genesis block is by definition in this state for all features.
```

Comment:
What happens if a fork happens and get back to the start height? Would the feature be evaluated again? Should we keep track of the feature state per block?

Reply:
Every feature is evaluated dynamically every time its state is queried on a specific block, by checking the state of that feature on all blocks up to the genesis. To make this efficient, every N blocks the state is cached, where N is the evaluation period time interval, defined in number of blocks.

When forks happen, the cache must be invalidated for every voided block. By doing that, the cache is preserved for blocks that weren't changed, and the new feature state is recalculated from the common block downwards.

Reference sections for this are "Caching mechanism", "Dealing with reorgs", and "Example".

Reply:
Got it. I didn't understand how the cache would work. Maybe I should read this section again. One suggestion might be to store information in blocks metadata. For instance, you might keep a dict of features and states. In this case, each block would copy this metadata from its parent and update according to its own bits. What do you think?

Reply:
That's a great suggestion, using a new metadata attribute to store states prevents the need of creating a new caching structure. But anyway, the caching logic remains mostly the same: when to set, retrieve, and remove the states from the metadata attribute.

Updated in https://github.com/HathorNetwork/rfcs/pull/51/commits/1c874a535200492260e8b5deb5c837df61629c11

Reply:
Great! Thanks!